### DISCUSSION TITLE
Rewrite url after `data` hook

### DISCUSSION THREAD

**Discussion by rtritto** [vikejs/vike#1843] [Discussion] [answered]
My use case is to rewrite (change/replace) the _url_ (without redirect) after some data are retrieved in the [data](https://vike.dev/data) hook.
How can it be done?

**Comment by brillout** [comment]
I presume you've already seen https://vike.dev/render?

**Reply by rtritto** [reply]
I did a try using `render` hook in `onBeforeRender` hook, the new page is rendered but the url path isn't updated.

---

Having the `/product/:name` route where _name_ can be a _name_ or an _id_.

My steps should be:
1. send request for `/product/:name` route where _name_ is an _id_
2. retrieve _data_ in `/pages/product/@name/+data`, find that _name parameter_ is an _id_ and then:
3. rewrite `/product/:id` url to `/product/:name` url 
4. render `/pages/product/@name/+Page`

Note: on rewrite of `/product/:id` url in `/product/:name` url, we must not pass again from `/product/:name` route because data (`/pages/product/@name/+data`) will be retrieved again

---

With `Next.js`, using `destination` and `permanent=false`, you can do:
- `/pages/product/[name]/index` file

  ```ts
  export async function getServerSideProps({ params: { name } }) {
    const data = getData(<SOME_FILTER>)
    if (isId(name) === true) {
      // name is an id
      return {
          redirect: {
            permanent: false,
            destination: `/product/${data.name}`
          },
          props: { data }
      }
    }
    return {
        props: { data }
    }
  }
  ```

**Reply by brillout** [reply]
> I did a try using `render` hook in `onBeforeRender` hook but didn't updated the url path.

I'd consider this a bug, reproduction welcome.

> Note: on rewrite of `/product/:id` url in `/product/:name` url, we must not pass again from `/product/:name` route because data (`/pages/product/@name/+data`) will be retrieved again

It's on the radar to be able to access the `pageContext` of previous render attempts. I created a feature request for it: https://github.com/vikejs/vike/issues/1846.

**Reply by rtritto** [reply]
> > I did a try using `render` hook in `onBeforeRender` hook but didn't updated the url path.
> 
> I'd consider this a bug, reproduction welcome.

Done, I opened #1847

**Comment by rtritto** [comment, marked as answer]
I fixed with:

```ts
// +Page.tsx

import { onMount } from 'solid-js'

export const Page = () => {
  onMount(() => {
    if (<SOME_CONDITION>) {
      // Update route path (no reload)
      globalThis.history.replaceState(null, '', <NEW_URL>)
    }
  })

  return (...)
}
```

Related to this use case, maybe #1846 isn't necessary to implement it.

FYI @brillout
